File: python/백준/그래프/BFS/2573 빙산.py
import sys
from collections import deque

input = sys.stdin.readline

N, M = map(int, input().split())
grid = [list(map(int, input().split())) for _ in range(N)]

directions = [(-1,0), (0,-1), (1,0), (0,1)]

def bfs(sx, sy, visited):
    queue = deque([(sx, sy)])
    visited[sx][sy] = True

    while queue:
        x, y = queue.popleft()

        for dx, dy in directions:
            nx, ny = x + dx, y + dy
            if 0 <= nx < N and 0 <= ny < M:
                if not visited[nx][ny] and grid[nx][ny] > 0:
                    visited[nx][ny] = True
                    queue.append((nx, ny))


def melt():
    # 각 칸이 얼마나 녹는지 먼저 계산
    minus = [[0] * M for _ in range(N)]

    for i in range(N):
        for j in range(M):
            if grid[i][j] > 0:
                sea_cnt = 0
                for dx, dy in directions:
                    ni, nj = i + dx, j + dy
                    if 0 <= ni < N and 0 <= nj < M:
                        if grid[ni][nj] == 0:
                            sea_cnt += 1
                minus[i][j] = sea_cnt

    # 실제로 녹이기
    for i in range(N):
        for j in range(M):
            if grid[i][j] > 0:
                grid[i][j] -= minus[i][j]
                if grid[i][j] < 0:
                    grid[i][j] = 0


year = 0

while True:
    # 1. 현재 빙산 덩어리 개수 세기
    visited = [[False] * M for _ in range(N)]
    count = 0

    for i in range(N):
        for j in range(M):
            if grid[i][j] > 0 and not visited[i][j]:
                bfs(i, j, visited)
                count += 1

    if count >= 2:
        print(year)
        break

    if count == 0:
        print(0)
        break

    # 2. 빙산 녹이기
    melt()
    year += 1

File: cpp/백준/그래프/BFS/2573 빙산.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

int N, M;
vector<vector<int>> grid;

void bfs(int sx, int sy, vector<vector<bool>>& visited) {
    int dx[] = {-1, 0, 1, 0};
    int dy[] = {0, -1, 0, 1};
    queue<pair<int, int>> q;
    q.push({sx, sy});
    visited[sx][sy] = true;

    while (!q.empty()) {
        auto [x, y] = q.front();
        q.pop();
        for (int i = 0; i < 4; ++i) {
            int nx = x + dx[i];
            int ny = y + dy[i];
            if (nx < 0 || nx >= N || ny < 0 || ny >= M) continue;
            if (!visited[nx][ny] && grid[nx][ny] > 0) {
                visited[nx][ny] = true;
                q.push({nx, ny});
            }
        }
    }
}

void melt() {
    int dx[] = {-1, 0, 1, 0};
    int dy[] = {0, -1, 0, 1};
    vector<vector<int>> minus(N, vector<int>(M, 0));
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < M; ++j) {
            if (grid[i][j] > 0) {
                int sea_cnt = 0;
                for (int k = 0; k < 4; ++k) {
                    int ni = i + dx[k];
                    int nj = j + dy[k];
                    if (ni < 0 || ni >= N || nj < 0 || nj >= M) continue;
                    if (grid[ni][nj] == 0) ++sea_cnt;
                }
                minus[i][j] = sea_cnt;
            }
        }
    }
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < M; ++j) {
            if (grid[i][j] > 0) {
                grid[i][j] -= minus[i][j];
                if (grid[i][j] < 0) grid[i][j] = 0;
            }
        }
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> N >> M;
    grid.assign(N, vector<int>(M));
    for (int i = 0; i < N; ++i)
        for (int j = 0; j < M; ++j)
            cin >> grid[i][j];

    int year = 0;
    while (true) {
        vector<vector<bool>> visited(N, vector<bool>(M, false));
        int count = 0;
        for (int i = 0; i < N; ++i) {
            for (int j = 0; j < M; ++j) {
                if (grid[i][j] > 0 && !visited[i][j]) {
                    bfs(i, j, visited);
                    ++count;
                }
            }
        }

        if (count >= 2) {
            cout << year << "\n";
            break;
        }
        if (count == 0) {
            cout << 0 << "\n";
            break;
        }

        melt();
        ++year;
    }
    return 0;
}
